### PR TITLE
Add support for php 8.1 and drop 7.2 and 7.3

### PR DESCRIPTION
**Pull request by marko-ilic**
In the scope of this PR, the support for PHP 8.1 is added (deprecated Serializable interface is removed). Since php 7.2 and 7.3 are no supported anymore, the support for them is removed in this PR.

### CODE REVIEW

**Comment by marko-ilic** [comment]
Hello @JeroenDeDauw 👋 
What do you think about changes in this PR?

**Comment by marko-ilic** [comment]
Hello @thiemowmde @manicki 
Can some of you, please, review this pr when you have time?

**Comment by thiemowmde** [comment]
I'm currently not active in the Wikidata project (which is where this library belongs). I have a few suggestions, though.
* The WMDE teams usually work with https://phabricator.wikimedia.org to track their resources. It might be a good idea to fill a ticket over there. That should make it easier for them to see this.
* From this request's description it's hard to tell how urgent it is. Note that Wikimedia servers are still using PHP 7 (for reasons nobody likes, but that's just how it is at the moment). Is there a ticket, sprint, team or something you can link this pull request to?
* The main commit in this patch is huge and does a lot of different things the same time. It would be better to split this into separate commits, possibly even separate pull requests. For example, the old array syntax still works in PHP 8.1, as far as I'm aware of. The patch also adds trailing commas. Why? I don't really see how such code style changes are related  to "support for PHP 8.1". Not doing them (or doing them separately) would make this patch much smaller and much easier to review.

**Comment by lucaswerkmeister** [comment]
>     * The WMDE teams usually work with https://phabricator.wikimedia.org to track their resources. It might be a good idea to fill a ticket over there. That should make it easier for them to see this.

I think this could fall under the existing task [T301249](https://phabricator.wikimedia.org/T301249). I’ll leave a comment there pointing to this PR.

I agree with Thiemo that this patch does too many things at once, and also that it’s too early to drop PHP 7.2 compatibility. Production is still on 7.2; in theory we could release Diff 4.0 requiring PHP 8.1+ now, and just keep using Diff 3.0 in Wikibase, but that means it’ll be a pain if we need any other Diff changes while still on PHP 7.2.

File: .travis.yml [hunk 1]
```diff
@@ -1,10 +1,9 @@
 language: php
-
+dist: bionic
 php:
-  - 7.2
-  - 7.3
   - 7.4
   - 8.0
+  - 8.1.0
```
Comment (reedy) [inline .travis.yml:6]: Do we need to pin to such a specific version?

**Comment by reedy** [comment]
I would concur too.

Adding PHP 8.1 support (especially if it's just `Serializable` related) can be done in a backwards compatible way.

File: .travis.yml (hunk 1 above)
Comment (reedy) [inline .travis.yml:6]: Do we need to pin to such a specific version?

Comment (marko-ilic) [reply .travis.yml:6]: > Do we need to pin to such a specific version?

This is Travis issue. It does not work if the version is `8.1` https://app.travis-ci.com/github/wmde/Diff/jobs/563853202

**Comment by manicki** [comment]
To keep things somehow focsued how about I take the topic to relevant WMDE teams (i.e. teams that would actually have to "deal with" the change), and we'll come back here, in particular with regards to the question of creating 8.1 supporting version, and the need/lack of thereof of adding it in a backwards compatible way? This very library had a similar conversation when PHP 7 enabled version was added before it could have been used in Wikibase on WMF servers. I think we can refer to what was learned then with regards to how did complicate things etc.

> Production is still on 7.2; in theory we could release Diff 4.0 requiring PHP 8.1+ now, and just keep using Diff 3.0 in Wikibase, but that means it’ll be a pain if we need any other Diff changes while still on PHP 7.2.

**Comment by marko-ilic** [comment]
> Adding PHP 8.1 support (especially if it's just `Serializable` related) can be done in a backwards compatible way.

There are also other deprecation warning e.g.

`Deprecated: Return type of Diff\DiffOp\Diff\Diff::offsetSet($index, $value) should either be compatible with ArrayObject::offsetSet(mixed $key, mixed $value): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /var/www/html/vendor/diff/diff/src/DiffOp/Diff/Diff.php on line 398`

`Deprecated: Return type of Diff\DiffOp\Diff\Diff::append($value) should either be compatible with ArrayObject::append(mixed $value): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /var/www/html/vendor/diff/diff/src/DiffOp/Diff/Diff.php on line 386`

**Comment by reedy** [comment]
The patches to MW core have been relatively trivial - https://github.com/wikimedia/mediawiki/commit/4cc1a3eecaaecc4368753a9a08fa0732bddae6fb

There hasn't been a release for 3.5 years - https://github.com/wmde/Diff/releases/tag/3.2.0

The point isn't just about Wikimedia servers though. You're going to want to have PHP 8.1 support (as people are using it), and if you can do that with minimal work, without maintaining support for multiple branches of the library (and then also allowing those in composer.json)...

The Serializeable related parts of the diff are pretty simple - https://github.com/wmde/Diff/pull/128/files#diff-ba4af8243477bbb63e951b5666ef96e257af4efd6e76c05c29f47563a9d5a0f8R65 for example

**Comment by marko-ilic** [comment]
I've prepared [a new PR](https://github.com/wmde/Diff/pull/129) with adding support for PHP 8.1 in a backward compatible way.

**Comment by manicki** [comment]
Thanks @marko-ilic for taking the effort to make the backwards-compatible approach

**Comment by reedy** [comment]
I think the other PR is ready for merge, other than maybe one minor issue with travis config (which may or may not have been fixed upstream). Waiting to see what https://github.com/wmde/Diff/pull/130 comes out with.

But can be squashed and merged as is; we can update the PHP 8.1 config in `.travis.yml` later easily enough

**Comment by lucaswerkmeister** [comment]
I think this has been superseded by #129 and other pull requests; please reopen if I missed something.

**Pull request closed without merging**